plugins/builds/index.js

```diff
@@ -765,7 +920,40 @@ const buildsPlugin = {
                             buildConfig.eventId = hoek.reach(pipelineJoinData[pid], 'event.id');
                         }
 
-                        if (buildConfig.eventId) {
+                        //   if nextBuild is stage teardown, just return nextBuild
+                        if (current.stage) {
+                            const stageDeletePromises = await handleStageFailure({
+                                nextJobName,
+                                current,
+                                buildConfig,
+                                jobFactory,
+                                buildFactory,
+                                username,
+                                scmContext
+                            });
+
+                            deletePromises.concat(stageDeletePromises);
+                        } else if (buildConfig.eventId) {
+                            const nextJobIsStageSetup = STAGE_SETUP_PATTERN.test(nextJobName);
+
+                            // Delete next stageBuild if next job is stage setup
+                            if (nextJobIsStageSetup) {
+                                const [, nextStageName] = nextJobName.match(STAGE_SETUP_PATTERN);
+                                const nextStage = await stageFactory.get({
+                                    name: nextStageName,
+                                    pipelineId: current.pipeline.id
+                                });
+
+                                if (nextStage) {
+                                    const nextStageBuild = await stageBuildFactory.get({
+                                        stageId: nextStage.id,
+                                        eventId: current.event.id
+                                    });
+
+                                    await nextStageBuild.remove();
```

Comment:
This should be abstracted into `build` model
https://github.com/screwdriver-cd/models/blob/master/lib/build.js#L652
```suggestion
                            const nextJobIsStageSetup = STAGE_SETUP_PATTERN.test(nextJobName);

                            // Delete next stageBuild if next job is stage setup
                            if (nextJobIsStageSetup) {
                                const [, nextStageName] = nextJobName.match(STAGE_SETUP_PATTERN);
                                const nextStage = await stageFactory.get({
                                    name: nextStageName,
                                    pipelineId: current.pipeline.id
                                });

                                if (nextStage) {
                                    const nextStageBuild = await stageBuildFactory.get({
                                        stageId: nextStage.id,
                                        eventId: current.event.id
                                    });

                                    await nextStageBuild.remove();
```

Reply:
https://github.com/screwdriver-cd/models/pull/606